### PR TITLE
Parsing generic subscripts

### PR DESCRIPTION
**Pull request by slavapestov**
Now that https://github.com/apple/swift/pull/7683 has landed, this adds parser support. The only tricky thing here is that since accessors for a subscript are not nested inside the subscript itself, we have to clone the generic parameter list of the subscript for each accessor, since otherwise name lookup won't find the generic parameters.

The main thing that remains now is SILGen support. Some simple examples I tried work, but I need to exercise the corner cases more, in particular materializeForSet, and actually write proper tests.

### CODE REVIEW

File: lib/Parse/ParseDecl.cpp [hunk 1]
```diff
@@ -5438,19 +5452,32 @@ Parser::parseDeclSubscript(ParseDeclOptions Flags,
...
                                                 CurDeclContext,
-                                                /*GenericParams=*/nullptr);
+                                                GenericParams);
   Subscript->getAttrs() = Attributes;
-  
+
+  // Code completion for the generic type params.
+  //
+  // FIXME: What is this?
+  if (GPHasCodeCompletion)
```
Comment (slavapestov) [inline lib/Parse/ParseDecl.cpp]: @nkcsgexi How do I test these 'code completion' code paths in the parser correctly?

File: lib/Sema/TypeCheckDecl.cpp [hunk 2]
```diff
@@ -4813,6 +4813,17 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
       }
     }
 
+    #if 0
```
Comment (DougGregor) [inline lib/Sema/TypeCheckDecl.cpp]: seems like we'll need this?

**Review by DougGregor — COMMENTED** [review]
Good start!

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Pull request merged**